### PR TITLE
[ML] DF Analytics should always display operational stats

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit populates the _stats API response with sensible "empty"
`data_counts` and `memory_usage` objects when the job itself
has not started reporting them.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java [hunk 1]
```diff
@@ -310,6 +318,31 @@ private static void parseHit(SearchHit hit, String configId, RetrievedStatsHolde
         );
     }
 
+    private List<Stats> buildStatsWithDefaults(List<DataFrameAnalyticsConfig> configs, List<Stats> stats) {
+        Map<String, DataFrameAnalyticsConfig> configById = new HashMap<>();
+        for (DataFrameAnalyticsConfig config : configs) {
+            configById.put(config.getId(), config);
+        }
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java]:
```suggestion
        Map<String, DataFrameAnalyticsConfig> configById = configs.stream().collect(Collectors.toMap(DataFrameAnalyticsConfig::getId, Function.identity()));
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java [hunk 2]
```diff
@@ -310,6 +318,31 @@ private static void parseHit(SearchHit hit, String configId, RetrievedStatsHolde
...
+
+        List<Stats> statsWithDefaults = new ArrayList<>(stats.size());
+        for (Stats statsItem : stats) {
+            DataFrameAnalyticsConfig config = configById.get(statsItem.getId());
+            statsWithDefaults.add(new Stats(
+                statsItem.getId(),
+                statsItem.getState(),
+                statsItem.getFailureReason(),
+                statsItem.getProgress(),
+                statsItem.getDataCounts() == null ? new DataCounts(config.getId()) : statsItem.getDataCounts(),
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java]: Why can't this and the memory usage be initialized in the stats ctor? 
```
this.memoryUsage = memoryUsage == null ? new MemoryUsage(config.getId(), Instant.now(), 0) : memoryUsage;
this.dataCounts = dataCounts == null ? new DataCounts(config.getId()) : dataCounts;
```

Why do we need to set `memory_usage` create time to the config's create time? Is that adding value for the added complexity?

Or maybe I am misunderstanding the `Stats#getId()` value...

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java]: I agree that's too much complexity for the gain. I'll make `MemoryUsage.timestamp` nullable.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java [hunk 3]
```diff
@@ -1507,7 +1508,7 @@ public void testGetDataFrameAnalyticsStats() throws Exception {
             .setAnalysis(OutlierDetection.createDefault())
             .build();
 
-        execute(
+        PutDataFrameAnalyticsResponse putConfigResponse = execute(
```
Comment (przemekwitek) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: Is this variable ever used?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/stats/MemoryUsage.java [hunk 4]
```diff
@@ -105,6 +119,7 @@ public String toString() {
     }
 
     public String documentId(String jobId) {
+        assert timestamp != null;
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/stats/MemoryUsage.java]: What prevents `timestamp` from being `null` here?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java [hunk 5]
```diff
@@ -42,10 +42,10 @@
 import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsState;
 import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsTaskState;
 import org.elasticsearch.xpack.core.ml.dataframe.stats.AnalysisStats;
-import org.elasticsearch.xpack.core.ml.dataframe.stats.common.DataCounts;
 import org.elasticsearch.xpack.core.ml.dataframe.stats.Fields;
 import org.elasticsearch.xpack.core.ml.dataframe.stats.MemoryUsage;
 import org.elasticsearch.xpack.core.ml.dataframe.stats.classification.ClassificationStats;
+import org.elasticsearch.xpack.core.ml.dataframe.stats.common.DataCounts;
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java]: Unused import?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java (hunk 3 above)
Comment (dimitris-athanasiou) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: Well spotted. I ended up not using it and forgot to clean it up.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/stats/MemoryUsage.java (hunk 4 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/stats/MemoryUsage.java]: We should only persist a `MemoryUsage` object we read from `AnalyticsResultProcessor` and those should always have a non-null timestamp.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java (hunk 5 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java]: It's used in line 206

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Review by przemekwitek — COMMENTED** [review]
LGTM

**Pull request merged**